### PR TITLE
Keep tooltip open when it gets hovered

### LINKED ISSUE #1131: Tooltip should remain open when the user hovers over the Tooltip itself
## 🐛 Bug Report 

Observed on https://react-spectrum.adobe.com/react-spectrum/Tooltip.html.

Tooltip closes when the user hovers over the Tooltip itself.

## 🤔 Expected Behavior

Tooltip should remain open when the user hovers over the Tooltip itself.

## 😯 Current Behavior

Tooltip closes when the user hovers over the Tooltip itself.

## 🔦 Context

Observed on https://react-spectrum.adobe.com/react-spectrum/Tooltip.html.

This is necessary to comply with [WCAG 2.1 Success Criterion 1.4.13: Content on Hover or Focus](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html), which states:

> Where receiving and then removing pointer hover or keyboard focus triggers additional content to become visible and then hidden, the following are true:
>
> **Dismissible**
> A mechanism is available to dismiss the additional content without moving pointer hover or keyboard focus, unless the additional content communicates an input error or does not obscure or replace other content;
>
> **Hoverable**
> _**If pointer hover can trigger the additional content, then the pointer can be moved over the additional content without the additional content disappearing;**_
>
> **Persistent**
> The additional content remains visible until the hover or focus trigger is removed, the user dismisses it, or its information is no longer valid.
>
>Exception: The visual presentation of the additional content is controlled by the user agent and is not modified by the author.

## 💻 Code Sample
```jsx
<TooltipTrigger>
  <ActionButton aria-label="Edit Name">
    <Edit />
  </ActionButton>
  <Tooltip>Change Name</Tooltip>
</TooltipTrigger>
```
## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  3.4.0
| Browser          | All
| Operating System | All

## 🧢 Your Company/Team

Adobe/Accessibility

### PR DESCRIPTION
**Pull request by Andarist**
Closes #1131 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

`yarn test`

### CODE REVIEW

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (Andarist) [inline packages/@react-aria/tooltip/test/useTooltip.test.js:27]: Having to do that (or using `useFocusVisible` which also leads to `setupGlobalFocusEvents`) is IMHO a bug - but it's a separate issue so I think it would be best to create a ticket for it and handle it later.

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (Andarist) [inline packages/@react-aria/tooltip/test/useTooltip.test.js:45]: OTOH having to do this sounds OK - I've scratched my head a little bit before I've realized that I have to do that (because `mouseenter` is not handled within `setupGlobalFocusEvents`) but in the real apps `mousemove` will always occur before `mouseenter` on the trigger element so I don't think that changing the implementation to cover testing quirks is required.

File: packages/@react-stately/tooltip/src/useTooltipTriggerState.ts
Comment (Andarist) [inline packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:44]: all options are optional so IMHO it makes sense to make `props` optional as well

File: packages/@react-stately/tooltip/src/useTooltipTriggerState.ts
Comment (Andarist) [inline packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:48]: this looks quirky but the return type of `setTimeout` is depending on env and on types loaded for that env so this ensures that it will always work~, regardless of what you are going to put into `tsconfig.json#compilerOptions.lib`

File: packages/@react-stately/tooltip/src/useTooltipTriggerState.ts
Comment (Andarist) [inline packages/@react-stately/tooltip/src/useTooltipTriggerState.ts:85]: I've chosen to reuse `TOOLTIP_COOLDOWN` which is equal to 500 but it might be worth shortening this timeout here, thoughts?

File: packages/@react-aria/tooltip/src/useTooltip.ts
Comment (Andarist) [inline packages/@react-aria/tooltip/src/useTooltip.ts:29]: I've chosen to accept `TooltipTriggerState` here because it contains everything that we need here. I couldn't make it required right now as it would be a breaking change but I believe that this should be done in the future.

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (Andarist) [reply packages/@react-aria/tooltip/test/useTooltip.test.js:27]: Filed an issue here: https://github.com/adobe/react-spectrum/issues/1378

File: packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js
Comment (Andarist) [inline packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js:489]: this test doesn't make a lot of sense now unless you'd like to specify different global and local cooldowns

**Comment by dannify** [comment]
Thanks for the contribution @Andarist! We will take a look.

**Review by snowystinger — COMMENTED** [review]
Hey! Thanks for starting this, I see the delay that's been added to closing. However, it doesn't look like the state has been passed through in the aria docs nor the react spectrum component, so it doesn't actually stay open on hover of the tooltip. Is there a missing commit? or was this the intent?

**Comment by Andarist** [comment]
@snowystinger I've focused on implementing `@react-aria` part as `@react-spectrum` is out of the scope of my interests 😉 It wasn't really too complicated so I've added this just now - please take a look. As to the docs - I've also added this to the existing example, not sure if this is enough or you'd like to add more information about this.

**Comment by snowystinger** [comment]
@Andarist That's understandable, I just wanted to make sure that was your intent. I appreciate that you added it to the other places, makes it easier to play with a live example.

File: packages/@react-aria/tooltip/docs/useTooltipTrigger.mdx
Comment (snowystinger) [inline packages/@react-aria/tooltip/docs/useTooltipTrigger.mdx:115]: I think you meant to add this to `Tooltip`?

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (snowystinger) [reply packages/@react-aria/tooltip/test/useTooltip.test.js:45]: yeah, might be good to have a util for this eventually, but we don't need to do it here

File: packages/@react-stately/tooltip/src/useTooltipTriggerState.ts
Comment (snowystinger) [reply packages/@react-stately/tooltip/src/useTooltipTriggerState.ts]: I think we can probably make it a prop to pass to useTooltipTriggerState if other libraries need to control that, but we can start with this default. For the record, it's not specified in Spectrum either right now. https://spectrum.adobe.com/page/tooltip/#Behaviors

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (snowystinger) [inline packages/@react-aria/tooltip/test/useTooltip.test.js:88]: we should pass `state` to `useTooltip` given your comment about this should be required? same comment for the other instances in this file
or is this to make sure that it works both ways?

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (snowystinger) [inline packages/@react-aria/tooltip/test/useTooltip.test.js:92]:
```suggestion
          <span {...triggerProps}>Trigger</span>
```

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (snowystinger) [inline packages/@react-aria/tooltip/test/useTooltip.test.js:93]:
```suggestion
          {state.isOpen && <span {...finalTooltipProps}>Tooltip</span>}
```

File: packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js
Comment (snowystinger) [reply packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js:489]: Maybe we change this test to use the `ESC` key to close it faster so we can check the cooldown

File: packages/@react-aria/tooltip/docs/useTooltipTrigger.mdx
Comment (Andarist) [reply packages/@react-aria/tooltip/docs/useTooltipTrigger.mdx]: yeah, this was done in a hurry 🤦 fixed

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (Andarist) [reply packages/@react-aria/tooltip/test/useTooltip.test.js:88]: > or is this to make sure that it works both ways?

This. As mentioned in some other comment - I think that this should be made required in the future but that would be a breaking change. We could add a single test ensuring that though and just make the rest use `state` - I'm fine either way, just let me know what do you think is best here.

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (Andarist) [reply packages/@react-aria/tooltip/test/useTooltip.test.js]: Good call - I've refactored this to use a button as a trigger element and this makes it easier to query it by role.

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (Andarist) [reply packages/@react-aria/tooltip/test/useTooltip.test.js]: Changed all occurrences of this pattern in a file 👍

File: packages/@react-aria/tooltip/test/useTooltip.test.js
Comment (snowystinger) [reply packages/@react-aria/tooltip/test/useTooltip.test.js:86]: I think we can leave as is then, I just wanted to make sure it was intentional.

File: packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js
Comment (Andarist) [reply packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js]: good idea - I've just refactored this test using the advice 👍

**Review by snowystinger — APPROVED** [review]
LGTM, I'll let the team know it's ready for its second review

File: packages/@react-aria/tooltip/src/useTooltip.ts
Comment (devongovett) [reply packages/@react-aria/tooltip/src/useTooltip.ts:29]: This is fine I think. Maybe we should add a `console.warn` here in case `state` isn't passed? That way it's not technically breaking but people know to update their code.

**Review by devongovett — APPROVED** [review]

**Review by ktabors — APPROVED** [review]
LGTM

File: packages/@react-aria/tooltip/src/useTooltip.ts
Comment (Andarist) [reply packages/@react-aria/tooltip/src/useTooltip.ts:29]: Yeah - this could be done. I was wondering - any particular reason why you have not been using `process.env.NODE_ENV` checks in the codebase so far? Such a warning would be a great candidate for a dev-only branch of code.

**Comment by Andarist** [comment]
One of the last thoughts on this one - if the same amount of time is used for both keeping the tooltip open after leaving the trigger and for the global "cooldown" then how useful the cooldown mechanism is? It still makes a functional difference for situations when a tooltip gets closed with ESC as noticed by @snowystinger but as the whole cooldown mechanism is focused on pointer interactions I'm not sure if this situation is really important - I would be inclined to say that no, it isn't. If you agree with that then the implementation could be simplified slightly by dropping that mechanism and making the warmup to rely on the module state of opened tooltips.

**Comment by snowystinger** [comment]
@Andarist Cooldown is different from time allowed to get hovered over the tooltip itselt. Cooldown is global to all tooltips, so if you hover one, it pops up, now if you hover a different target, the tooltip will appear instantly. There's a series of stories with multiple tooltips where you can try this out if you like, it's also explained in our documentation here https://react-spectrum.adobe.com/react-spectrum/Tooltip.html#warmup--cooldown

**Comment by Andarist** [comment]
I understand this mechanism - I might have not described my intention too well though so let me rephrase.

I believe that the cooldown could be removed because when leaving a trigger a tooltip doesn't disappear immediately (unlike before the changes from this PR). The disappearing of the tooltip could act as a cooldown mechanism - if we hover another trigger element when the previous tooltip is still displayed then we switch to displaying the new tooltip anyway (as the mechanism has been warmed up already) and if the tooltip disappears then we could consider this as cooldown delay passed so if we hover over whatever trigger after that then we'll have to wait for a new warmup.

**Comment by snowystinger** [comment]
Ahhhh, yes, I misunderstood. Thank you for clarifying. I'll ask Design, though we may not get an answer back until the new year.

**Comment by Andarist** [comment]
Sure, no problem. This doesn't have to block this PR though - it's just a potential refactor opportunity for the future. Do you think that those 500ms is the appropriate time for keeping the tooltip open when leaving the trigger?

**Comment by snowystinger** [comment]
Yep, 500ms is fine for this PR

**Pull request merged**
